### DISCUSSION TITLE
ODD Behavior of CORE emulator for saving configuration of nodes

### DISCUSSION THREAD

**Discussion by abdul152** [coreemu/core#538] [Q&A]
Hi guys, 
I need one help I am Using CORE emulator on Ubuntu 20,04 LTS.  I am also using FRRouting kernel with it.  I read the documentation of frr and CORE as well and as per the expection it seems work normal. Like If I add nodes on CORE and configure them with frr using FRRospfv2 and FRRzebra. these are working fine. If these nodes are working with frr vtysh and I configured them with "service integrated-vtysh-config" then all of these nodes saving the configuration in one single file at /etc/frr/frr.conf.  see the output below,

Hello, this is FRRouting (version 7.2.1).
Copyright 1996-2005 Kunihiro Ishiguro, et al.

n1# configure t
n1(config)# service integrated-vtysh-config 
n1(config)# do wr
Note: this version of vtysh never writes vtysh.conf

Warning: attempting direct configuration write without watchfrr.
File permissions and ownership may be incorrect, or write may fail.

Building Configuration...
Integrated configuration saved to /etc/frr/frr.conf
[OK]
n1(config)# 

But, If I disbale this "service integrated-vtysh-config" this from nodes by this command  "n1(config)# no service integrated-vtysh-config, then try to save the configuration so, I got an error please see below,

Hello, this is FRRouting (version 7.2.1).
Copyright 1996-2005 Kunihiro Ishiguro, et al.

n2# configure t  
n2(config)# no service integrated-vtysh-config 
n2(config)# do wr
Note: this version of vtysh never writes vtysh.conf
Building Configuration...
Can't backup old configuration file /usr/local/etc/frr/frr.conf.sav.
Can't backup old configuration file /usr/local/etc/frr/frr.conf.sav.
n2(config)# 

After issuing the command on nodes  " no service integrated-vtysh-config" nodes change its directory to save configuration and its getting failed to save it.

This is the behavious of FRR with CORE. 

Now another thing is that, 
If on the same Ubuntu vm If I go to host terminal instead of Node terminal on CORE. and run vtysh command so it opened one frrouting instace and on that window I can configure it as well. and on that window if I use "service integrated-vtysh-config" then all the configuration of that instace is saving on /etc/frr/frr.conf and if I run no service integrated-vtysh-config command then all the configuration is going to save on same directory but with differnt files for every procol like ospf, zebra, isis, and bgp 
It is like you can see below, 

abdul@COREFRR:~/Desktop$ sudo vtysh 

Hello, this is FRRouting (version 7.2.1).
Copyright 1996-2005 Kunihiro Ishiguro, et al.

COREFRR# wr
Note: this version of vtysh never writes vtysh.conf
Building Configuration...
Integrated configuration saved to /etc/frr/frr.conf
[OK]
COREFRR# configure t
COREFRR(config)# no service integrated-vtysh-config 
COREFRR(config)# do wr
Note: this version of vtysh never writes vtysh.conf
Building Configuration...
Configuration saved to /etc/frr/zebra.conf
Configuration saved to /etc/frr/ospfd.conf
Configuration saved to /etc/frr/bgpd.conf
Configuration saved to /etc/frr/staticd.conf
COREFRR(config)# 

Please help me out for this purpose. 
Thanks you so much in advance.

**Comment by bharnden** [comment]
I am not familiar with your intended goal or FRR to answer your questions. They maybe better suited for FRR developers or their documentation.

The one thing you have to keep in mind for CORE is the following, which given known knowledge of how things work in general, the nuance of using it within a CORE node.

- CORE nodes share the same host file system as the host itself
- They make some isolation of their generated files by way of using a node specific directory `/tmp/pycore.<session id>/<node name>.conf`
- Running software that depends on using file paths outside of that on multiple nodes would cause issues due to collision, ie `/etc/frr/frr.conf` would be the same across all nodes, unless a CORE service is written to attempt to compensate for this
- This is why some CORE services specify directories to mount unique to themselves, to avoid this collision

It seems like you are running into issues with FRR attempting to use paths that are non unique to hosts and how the software interacts with those paths and files. The FRR services seems to try and make some compensations, in particular if `/etc/frr/frr.conf` does not exist, but if you are running commands on the host and creating said file, you may be causing odd behavior.

That is about as much as I can tell you, unless someone more familiar with FRR specifics can chime in, otherwise your best bet would be further understanding how to use FRR itself and dealing with file pathing related nuances.
